### PR TITLE
Ripout state

### PR DESCRIPTION
**Pull request by martindurant**

### CODE REVIEW

**Comment by martindurant** [comment]
Replaces #61

**Comment by martindurant** [comment]
@seibert , you may find this interesting. Single file and multi-file cats now actual data-source plugins with type "catalog".

**Comment by martindurant** [comment]
So, states are gone, plugins are gone and containers are gone. This would require bumping versions on all existing plugins.

File: intake/catalog/local.py [hunk 1]
```diff
@@ -120,21 +118,19 @@ def __init__(self, name, description, driver, direct_access, args,
...
-    def find_plugin(self, registry):
-        if self._driver in registry:
-            self._plugin = registry[self._driver]
-        else:
-            self._plugin = global_registry[self._driver]
-
     def describe(self):
+        if isinstance(self._plugin.container, (list, tuple)):
+            import pdb
+            pdb.set_trace()
```
Comment (mmccarty) [inline intake/catalog/local.py]: Don't forget to remove

File: intake/catalog/local.py [hunk 2]
```diff
@@ -120,21 +118,19 @@ def __init__(self, name, description, driver, direct_access, args,
...
+            print('####', self._plugin)
```
Comment (mmccarty) [inline intake/catalog/local.py]: Perhaps we should have an issue to come back and replace `print` with a logger?

File: intake/catalog/local.py [hunk 3]
```diff
@@ -326,18 +261,25 @@ def _parse_plugins(self, data):
                 self.error("keys 'module' and 'dir' both exist (select only "
                            "one)", plugin_source)
             elif 'module' in plugin_source:
-                obj = self._parse_plugin(plugin_source, 'module')
-                if obj:
-                    sources.append(obj)
+                for k, v in load_plugins_from_module(
```
Comment (mmccarty) [inline intake/catalog/local.py]: For this block and the one below, it feels like there's enough here to warrant a separate function (`register_module` and `register_dir`).

File: intake/container/base.py [hunk 4]
```diff
@@ -1,21 +1,81 @@
...
-    @staticmethod
-    def decode(bytestr):
-        raise NotImplementedError
+    def _get_source_id(self):
+        if self._source_id is None:
+            payload = dict(action='open', name=self.name,
+                           parameters=self.parameters)
+            req = requests.post(self.url, data=msgpack.packb(
+                payload, use_bin_type=True), **self.headers)
+            if req.status_code == 200:
```
Comment (mmccarty) [inline intake/container/base.py]: Do we need to raise an exception here?  We do that above.

File: intake/container/base.py [hunk 5]
```diff
@@ -1,21 +1,81 @@
...
+        self.npartitions = response['npartitions']
+        self.metadata = response['metadata']
+        self._schema = Schema(datashape=None, dtype=self.dtype,
+                              shape=self.shape,
+                              npartitions=self.npartitions,
+                              extra_metadata=self.metadata)
+        self._source_id = response['source_id']
+
+    def _get_partition(self, i):
+        pass
```
Comment (mmccarty) [inline intake/container/base.py]: todo?

File: intake/source/base.py [hunk 6]
```diff
@@ -139,14 +99,7 @@ def discover(self):
...
         """Load entire dataset into a container and return it"""
-        self._load_metadata()
-
-        parts = [self._get_partition(i) for i in range(self.npartitions)]
-
-        return self._merge(parts)
-
-    def _merge(self, parts):
-        return get_container_klass(self.container).merge(parts)
+        pass
```
Comment (mmccarty) [inline intake/source/base.py]: todo?

File: intake/__init__.py [hunk 7]
```diff
@@ -60,6 +69,44 @@ def output_notebook(inline=True, logo=False):
...
+                else:
+                    # single filename in a list
+                    driver = 'yaml_file_cat'
+            elif isinstance(uri, str):
+                # single URL
+                if uri.startswith('intake:'):
+                    # server
+                    driver = 'intake_remote'
+                else:
+                    if uri.endswith('.yml') or uri.endswith('.yaml'):
```
[the remark this answers is not in the document]
Comment (martindurant) [reply intake/__init__.py]: Always forget about that

File: intake/catalog/local.py (hunk 1 above)
Comment (martindurant) [reply intake/catalog/local.py]: Doesn't actually get called. Was caused by one of the datasources having a rogue comma, making a string into a tuple.

File: intake/container/base.py (hunk 5 above)
Comment (martindurant) [reply intake/container/base.py]: I prefer no-op over virtual methods or NotImplemented, just for simplicity.

File: intake/catalog/local.py (hunk 3 above)
Comment (martindurant) [reply intake/catalog/local.py]: done

File: intake/container/base.py (hunk 4 above)
Comment (martindurant) [reply intake/container/base.py]: done

File: intake/catalog/local.py (hunk 2 above)
Comment (martindurant) [reply intake/catalog/local.py]: done

File: intake/catalog/local.py (hunk 1 above)
Comment (martindurant) [reply intake/catalog/local.py]: done

File: intake/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (martindurant) [reply intake/__init__.py]: done

File: intake/container/base.py (hunk 5 above)
Comment (mmccarty) [reply intake/container/base.py]: Perhaps a comment then to explain why this is not implemented.

Comment (martindurant) [reply intake/container/base.py]: Do you personally prefer NotImplemented? That's what's done in the DataSource base class.

Comment (mmccarty) [reply intake/container/base.py]: It depends.  I think the noop is fine if a noop is the default.  If you are depending on the method to actually do something, then I would raise NotImplemented.

Comment (martindurant) [reply intake/container/base.py]: done

File: intake/source/base.py (hunk 6 above)
Comment (martindurant) [reply intake/source/base.py]: done

**Comment by martindurant** [comment]
Relented... those methods should never be directly called, all containers should override them.

Also added a nested test :)

**Comment by martindurant** [comment]
@seibert , remind me: if I merge this, it will build a <dev> package, and only tagging will result in a <main> package, correct?

**Comment by martindurant** [comment]
(the documentation will need updating before release, especially the "how to make a plugin", because we no longer need `Plugin`; also all existing plugins would need to be changed and released very shortly after intake, and depending on the new version)

**Comment by seibert** [comment]
This repo uploads every merge to master with a build number that increments for each commit since the most recent tag (basically `git describe`).  That is useful for testing, but we should be uploading them with a `dev` label.

**Comment by martindurant** [comment]
What would it take to get the dev/main label behaviour? I am *not* ready to start a release process!

**Comment by seibert** [comment]
I think it would require changing the anaconda upload line to set the label (and probably copying the bash if statement from one of the plugins to branch for tagged and untagged commits).

**Pull request merged**